### PR TITLE
Add classes for Coils and CoilSets

### LINKED ISSUE #90: Classes for Coils, Coil Groups, Coilsets etc
- [x] Coil - an individual coil
- [x] Coil Group - several coils that share the same current / parameters? IE same shape at different toroidal angles?
- [x] Coilset - all the coils needed to make a plasma

Each should have a `compute_magnetic_field` method and maybe also for vector potential?

### LINKED ISSUE #146: Allow specification of curve coordinate system
ie, all the ```compute_``` methods of curves that return a coordinate or vector should allow the user to specify whether it should be in cartesian or cylindrical coordinates

### LINKED ISSUE #147: Add class for rotated/shifted/reflected curves
ie, a curve that is rotated about phi, like in a coilset. Ideally should wrap multiple types of underlying curves, and able to handle different coordinate systems

### PR DESCRIPTION
**Pull request by f0uriest**
- Adds a base class for ``Coil`` that inherits from ``MagneticField`` and has ``compute_magnetic_field`` method
- Classes for specific coil parameterizations inherit from ``Coil`` and specific classes of ``Curve`` (eg, ``FourierRZCurve``, ``FourierPlanarCurve``)
- Adds ``CoilSet`` class that can wrap multiple coils (or other ``CoilSet``s) and compute the combined field from all coils.
- Adds ``basis`` option to ``Curve`` and ``MagneticField`` compute methods for specifying R, phi, Z or X, Y, Z basis for input coordinates and returned vectors.
- Add methods to ``translate``, ``rotate``, ``flip`` curves and coils for rigid displacements and rotations independent of parameterization.
- Utilities for converting between coordinate systems moved to ``geometry.utils``
- Increase resolution for ``Curve`` default grids
- Change default parameters for ``FourierXYZCurve`` to give posiitve toroidal field for positive current (consistent with ``FourierPlanarCurve``
- Remove computation of field derivatives for ``MagneticField`` classes - these were unused, and are hard to get to work intuitively with the choice of basis. May get added back later.

Resolves #90 
Resolves #146 
Resolves #147

### CODE REVIEW

**Comment by dpanici** [comment]
increased the default grid size in commit d2a5c7c due to the following:
`print('N | Magnetic Field at Center of Loop')`

`for ng in np.arange(0,50,5)[1:]:`
`    C = FourierXYZCoil(X_n=[0,10,2],Y_n=[2,0,0],Z_n=[0,0,0],modes=[-1,0,1])`
`    C.grid = LinearGrid(N=int(ng), endpoint=True)`
`    print(ng,C.compute_magnetic_field(Grid([[10,0,0]]),basis='xyz'))`

yields the following, where the analytic answer is B_z=pi e-07

> N | Magnetic Field at Center of Loop
> 5 [[0.e+00 0.e+00 4.e-07]]
> 10 [[0.00000000e+00 0.00000000e+00 3.27573211e-07]]
> 15 [[0.00000000e+00 0.00000000e+00 3.19540864e-07]]
> 20 [[0.00000000e+00 0.00000000e+00 3.17053924e-07]]
> 25 [[0.00000000e+00 0.00000000e+00 3.15965994e-07]]
> 30 [[0.00000000e+00 0.00000000e+00 3.15394007e-07]]
> 35 [[0.00000000e+00 0.00000000e+00 3.15056397e-07]]
> 40 [[0.00000000e+00 0.00000000e+00 3.14840549e-07]]
> 45 [[0.00000000e+00 0.00000000e+00 3.14694211e-07]]

So default resolution for simple coils was too low

**Review by dpanici — APPROVED** [review]
Looks good, made a minor change to default grid resolution for curves

File: desc/coils.py [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+            for coil in coils[::-1]:
+                fcoil = coil.copy()
+                fcoil.flip(normal)
+                fcoil.flip([0, 0, 1])
+                fcoil.current = -1 * coil.current
+                flipped_coils.append(fcoil)
+            coils = coils + flipped_coils
+        for k in range(0, NFP):
+            coil = coils.copy()
+            coil.rotate(axis=[0, 0, 1], angle=2 * jnp.pi * k / NFP)
```
Comment (ddudt) [inline desc/coils.py:418]: Won't the `k=0` iteration of this just duplicate the existing field period of coils without rotating them?

Comment (ddudt) [reply desc/coils.py:418]: never mind

**Review by ddudt — APPROVED** [review]
`CoilSet.from_symmetry` looks correct when plotted.

**Pull request merged**